### PR TITLE
Add runtime assertions for input shape constraints

### PR DESCRIPTION
**Pull request by tugsbayasgalan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #100247

This PR adds runtime assertions as an extra pass in the exported graph. Several high level information:
1. We specialize all dimensions that were not added to the user input constraints 
2. We haven't added relational constraints as runtime assertions (e.g x[1] == x[0]), will do in a follow up diff

Differential Revision: [D45408971](https://our.internmc.facebook.com/intern/diff/D45408971)

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45408971).

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45408971).

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45408971).

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45408971).

File: torch/_export/passes/add_runtime_assertions_for_constraints_pass.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+                f"specialized at {shapes[dim]}"
+            )
+            dim_node = self._fx(
+                "call_function",
+                torch.ops.aten.sym_size,
+                (arg, dim),
+                {},
+                self._create_dummy_node_metadata(current_fake_mode),
+                self.interpreter,
+            )
```
[the remark this answers is not in the document]
Comment (tugsbayasgalan) [reply torch/_export/passes/add_runtime_assertions_for_constraints_pass.py:106]: Yep!

File: torch/_export/passes/add_runtime_assertions_for_constraints_pass.py [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+
+ConstraintSpec = namedtuple("ConstraintSpec", ["constraint_dim", "min_val", "max_val"])
+
+
+class AddRuntimeAssertionsForConstraintsPass(ExportPassBase):
+    def __init__(self) -> None:
+        super().__init__()
+        self.current_gm = None
+
+    def _create_dummy_node_metadata(self, fake_mode, example_value: Optional[torch.Tensor] = None) -> NodeMetadata:
```
[the remark this answers is not in the document]
Comment (angelayi) [reply torch/_export/passes/add_runtime_assertions_for_constraints_pass.py:27]: Hmm, so technically all we do with the metadata is update the node in the graph to whatever metadata the user has inserted except for the protected keys (https://fburl.com/yaqgysim), and then we update the "val" metadata with the result of running that op/node (https://fburl.com/rofg11hc). So I think it shouldn't matter what you set here...

File: test/export/test_passes.py [hunk 3]
```diff
@@ -58,6 +68,109 @@ def count_additions(gm) -> int:
...
+
+        with self.assertRaisesRegex(RuntimeError, "Input #0"):
+            pass_result.graph_module(torch.zeros(4, 7, 3), torch.ones(5, 5, 5))
+
+        # y is specialized to 5
+        with self.assertRaisesRegex(RuntimeError, "Input #1's dimension #0 size is specialized at 5"):
+            pass_result.graph_module(torch.zeros(4, 2, 3), torch.ones(2, 5, 5))
+
+        # Since we didn't insert the constraint for x[1] >= 2, it should work for case where x[1] == 1
+        gm_result_for_1_size = pass_result.graph_module(torch.ones(3, 1, 3), torch.ones(5, 5, 5))
```
Comment (guangy10) [inline test/export/test_passes.py:169]: I think the contract is that export API will guarantee that user specified constraints are always respected (as long as valid). This seems like a flaw and it will be a surprise to users when `dynamic_dim(x, 1) >= 2` is specified but not enforced. I think the motivation of not insert assertion for `x[1] >= 2` is because of 0/1 specialization where in most cases, N >= 2 will work also work for N == 1. Is it the reason?

If yes maybe a better option is to let the compiler to figure out whether the exported graph at dim 1 can be correctly generalize to N == 1, which should be simple and clean to validate in a separate pass prior to insert runtime assertion. If N == 1 works correctly, remove constraint for `dynamic_dim(x, 1) >= 2` and warn users, or error out and let users remove the constraint from the input explicitly, the remaining input constraints will be correct and necessary.

File: torch/_export/passes/add_runtime_assertions_for_constraints_pass.py [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+                min_vals = [x[0] for x in constraints]
+                max_vals = [x[1] for x in constraints]
+                min_val = sorted(min_vals, reverse=True)[0]
+                max_val = sorted(max_vals, reverse=False)[0]
+
+                assert min_val <= max_val
+
+                constraints_id_to_constraint[t_id].append(
+                    ConstraintSpec(constraint_dim=dim, min_val=min_val, max_val=max_val)
+                )
```
Comment (guangy10) [inline torch/_export/passes/add_runtime_assertions_for_constraints_pass.py:67]: All the processing here is to reorganize the constraints by input tensor. I'm not sure why the `Constraint` class is not defined as a ref to a single tensor with a *list* of dims and ranges in the very first place.

File: torch/_export/passes/add_runtime_assertions_for_constraints_pass.py [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+
+        return constraints_id_to_constraint
+
+    def call(self, graph_module: torch.fx.GraphModule) -> None:
+        # resets the counter
+        self.input_tracker = 0
+        self.current_gm = graph_module
```
Comment (guangy10) [inline torch/_export/passes/add_runtime_assertions_for_constraints_pass.py:74]: Should make a copy? Any chance the original graph may be polluted?

File: torch/_export/passes/add_runtime_assertions_for_constraints_pass.py [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+            self._insert_specialized_shapes_assert(arg, all_dims, current_inp)
+            self.input_tracker += 1
+            return arg
+
+        constraints = self.constraints[id(current_inp)]
+
+        constrained_dims = set()
+        # Add runtime asserts for user specified constraints for each
+        # individual dimensions (e.g not the relational constraints like
+        # x[1] == x[0])
```
Comment (guangy10) [inline torch/_export/passes/add_runtime_assertions_for_constraints_pass.py:130]: This is not a TOOD right? IIUC even with relational constraints, we will assert on each dim independently?

File: torch/_export/passes/add_runtime_assertions_for_constraints_pass.py [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+                NodeMetadata({}),
+            )
+            tensor_eq = self.call_operator(
+                torch.ops.aten.scalar_tensor.default,
+                (eq,),
+                {},
+                NodeMetadata({}),
+            )
+            self.call_operator(
+                torch.ops.aten._assert_async.msg,
```
Comment (guangy10) [inline torch/_export/passes/add_runtime_assertions_for_constraints_pass.py:106]: Curious what this does: Is it inserting the assert op in the exported graph?

File: torch/_export/passes/add_runtime_assertions_for_constraints_pass.py [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+                (arg, constraint.constraint_dim),
+                {},
+                NodeMetadata({}),
+            )
+            assert_msg = (
+                f"Input #{self.input_tracker}'s dimension #{constraint.constraint_dim} size is "
+                f"outside of specified dynamic range [{constraint.min_val}, {constraint.max_val}]"
+            )
+
+            if constraint.min_val > 2:
```
Comment (guangy10) [inline torch/_export/passes/add_runtime_assertions_for_constraints_pass.py:144]: See my comment above for not inserting `dynamic_dim(t_x, dim_y) >= 2`

**Review by guangy10 — COMMENTED** [review]
Just to confirm, turn inline constraints to runtime assertion will be addressed in a separate PR?

File: torch/_export/passes/add_runtime_assertions_for_constraints_pass.py (hunk 7 above)
Comment (angelayi) [reply torch/_export/passes/add_runtime_assertions_for_constraints_pass.py:106]: it's inserting the assert op in a new graph that will be returned from running this pass.

File: torch/_export/passes/add_runtime_assertions_for_constraints_pass.py (hunk 6 above)
Comment (tugsbayasgalan) [reply torch/_export/passes/add_runtime_assertions_for_constraints_pass.py:130]: Yep we will. But I think we also should add relational constraints as well. If user has a constrain:
dynamic_dim(x, 0) >= 3, dynamic_dim(x, 1) >= 3, and dynamic_dim(x, 0) == dynamic_dim(x, 1), we still need to error for cases where x.shape[0] = 4 and x.shape[1] = 5.

File: test/export/test_passes.py (hunk 3 above)
Comment (tugsbayasgalan) [reply test/export/test_passes.py:169]: Yeah our assumption is that if we were able to get a graph for N >= 2, it should work for cases that N == 1. 
For the validation, I agree that we need an infra that can check if our graph can be correctly specialized for N==1 cases. Do you think we can do it in a follow up diff? 
cc: @avikchaudhuri

Comment (guangy10) [reply test/export/test_passes.py:169]: Yeah not mean to block. Add a comment/todo about the assumption, and follow up in a separate PR is okay.

The important thing is that I want to make sure this will be handled decently as part of export contract. Currently validating whether the graph can be correctly generalized to N == 1 or not is user's responsibility, e.g. they have to write tests for it. This is conflict to not exposing 0/1 specialization to end users, and what's worse when users realize that the exported graph can not be generalized to N == 1, they have no way to enforce the correctness because the constraint N >= 2 is not installed and there are no knobs for users to correct it manually. So IMO it's best to make it part of export contract. 

Like to hear your thoughts @avikchaudhuri

File: torch/_export/passes/add_runtime_assertions_for_constraints_pass.py (hunk 6 above)
Comment (guangy10) [reply torch/_export/passes/add_runtime_assertions_for_constraints_pass.py:130]: Yes, you're correct. I recall that Avik clarified what equality means in his PR. Ok so it is a TODO to support equality constraints

File: torch/_export/passes/add_runtime_assertions_for_constraints_pass.py (hunk 4 above)
Comment (guangy10) [reply torch/_export/passes/add_runtime_assertions_for_constraints_pass.py:67]: I'm totally fine with it.

**Review by guangy10 — APPROVED** [review]
Assume you will address linter error and other test failures if relevant. The PR itself looks good to me.

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45408971).

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45408971).

File: torch/_export/passes/add_runtime_assertions_for_constraints_pass.py [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+
+    def _insert_specialized_shapes_assert(self, arg, dims, current_inp):
+        # we don't want to get shapes from meta as they will be symbolic
+        shapes = current_inp.shape
+        for dim in dims:
+            assert_msg = (
+                f"Input #{self.input_tracker}'s dimension #{dim} size is "
+                f"specialized at {shapes[dim]}"
+            )
+            dim_node = self.call_operator(
```
Comment (ydwu4) [inline torch/_export/passes/add_runtime_assertions_for_constraints_pass.py:89]: Non blocking. Using call_operator here seems to make implementation of inline constraints more complicated: in the overriding call_operator function, we need to differentiate whether the call operator is from the original graph module or the newly inserted assertions. Can we use _fx to make it clear that these are the newly added nodes?

Comment (ydwu4) [reply torch/_export/passes/add_runtime_assertions_for_constraints_pass.py:89]: Nvm, let's just test whether meta is empty or not to differentiate.

File: test/export/test_passes.py (hunk 3 above)
Comment (avikchaudhuri) [reply test/export/test_passes.py:169]: As discussed in the meeting yesterday, there are two cases:
* If export passes with N >= 2, we only have a soft expectation that it should also work with N == 1. This can only be covered by testing, I don't think we can validate this automatically in general. It also feels wrong to disallow N == 1 in that case. So I suggest we just ask people to test if they think that N == 1 should be a valid input.
* If you try exporting with N == 1, we should add a `log.warning` that we are implicitly specializing. If you marked that dimension dynamic, we should error.

**Review by avikchaudhuri — APPROVED** [review]

File: test/export/test_passes.py (hunk 3 above)
Comment (guangy10) [reply test/export/test_passes.py:169]: @avikchaudhuri I agree with #2 _(warning when exporting with N == 1)_, I think #1 needs more discussion.

> we only have a soft expectation that it should also work with N == 1

That's the problem because soft expectation means no guarantee at all and hope it works. When it doesn't work, it will come as a surprise to users, as they have specified `dynamic_dim(x, 1) >= 2` as an input constraint based on the suggestion provided by the compiler! The promise of user specified constraints will be respected faithfully is no longer true in this case.

> This can only be covered by testing, I don't think we can validate this automatically in general.

I fail to understand why this can not be validated automatically in general if this is testable by users. Maybe you could help elaborate why you think the approach I proposed in previous comment may not work? We can brainstorm other solutions when we agree this is an issue we need to take care for the users.
 
 > It also feels wrong to disallow N == 1 in that case. So I suggest we just ask people to test if they think that N == 1 should be a valid input.
 
I'm not suggesting whether N == 1 should be allowed or not allowed. If we can reason about either way I will be fine. My point is that the contract provided by export should be cleaner and stronger, at least on input shape constraints.

Comment (avikchaudhuri) [reply test/export/test_passes.py:169]: OK, so we are only talking about you passing in an example input where N >= 2 and we pass, right? Now we have only two options:
1. Assert that N >= 2. I don't like this option because in practice your code might be fine with N == 1, and you can't export with N == 1 and dynamic, so you're stuck if you also wanted N == 1 to work.
2. Automatically validate. This seems tricky. Will we come up with new example inputs with N == 1? Or basically use fake tensors directly? Will we try all combinatorial possibilities for all dynamic inputs, setting each N to 1? Etc. It is because of this trickiness that I want to just suggest that users do testing for this case.

Does this make sense? Please let me know if I'm missing anything easy here.

> When it doesn't work, it will come as a surprise to users, as they have specified dynamic_dim(x, 1) >= 2 as an input constraint based on the suggestion provided by the compiler! The promise of user specified constraints will be respected faithfully is no longer true in this case.

I didn't understand this. If you specify dynamic_dim(x, 1) >= 2, and if it doesn't work with N == 1 then it's fine! So we're only talking about when you specify dynamic_dim(x, 1) without bounds.

Comment (guangy10) [reply test/export/test_passes.py:169]: @avikchaudhuri Sorry for late reply. I want to emphasize that my point is that I feel the export contract should be stronger with 0/1 specialization.

> I didn't understand this. If you specify dynamic_dim(x, 1) >= 2, and if it doesn't work with N == 1 then it's fine!

It's fine only if it happens to work for N == 1 because users won't get the wrong result. It's not fine if it doesn't work for `N == 1`. I think this is the key point we haven't aligned yet. 
For example, say I have `2 <= dynamic_dim(x, 1) <= 10`, if I give an input `x.size(1) == 11`, it errors out immediately right? Even if `x.size(1) == 11` may happen to work, we want to error out anyway because that's the constraint user specified. Why is `x.size(1) == 1` treated differently? 


> Now we have only two options:

I don't see option 1 & 2 are exclusive. We probably should assert `N >= 2` conditionally. So let's talk about option 2 first. If there is a way to validate whether or not the exported graph (for `N >= 2` at all dynamic dims) works for `N == 1`, then we can assert `N >= 2` conditionally. Actually there is not much dims to test against. It's linear. If you have an input tensor of size `(x, y, z)`, in the worst case where all dims are dynamic, you generate `x + y + z` inputs in total that toggle one dim to 1 a time to test if the graph can be correctly generalized to N == 1, you only install assertion of `N >= 2` on dim where `N == 1` produces inconsistent output.

Comment (avikchaudhuri) [reply test/export/test_passes.py:169]: > For example, say I have 2 <= dynamic_dim(x, 1) <= 10, if I give an input x.size(1) == 11, it errors out immediately right? Even if x.size(1) == 11 may happen to work, we want to error out anyway because that's the constraint user specified. Why is x.size(1) == 1 treated differently?

It is treated differently because PyTorch treats it differently. I'm assuming there is a good reason for 0/1 specialization: I'm guessing it is because many torch operators behave differently for 0/1 than others. This is something I'm willing to defer to and explain away based on general PyTorch semantics.

**Is it ideal? NO. So I fully agree with you on the jankiness here.** But at the same time I'm not comfortable challenging the assumption that 0/1 specialization predates export and exists for a good reason, and taking a hard stand with export without understanding all the reasoning behind why it exists.

> Actually there is not much dims to test against. It's linear.

This is an incomplete test. For each dynamic, we have two cases, 1 or >= 2. So a complete test is exponential: 2^N. You're missing the case where x = 1, y = 1, z >= 2, e.g.

I won't oppose doing the incomplete linear test to surface additional signal.
